File: game-server/src/main/java/org/forfun/mmorpg/game/database/user/entity/PlayerEnt.java
package org.forfun.mmorpg.game.database.user.entity;

import jakarta.persistence.Convert;
import jakarta.persistence.Transient;
import org.forfun.mmorpg.game.base.GameContext;
import org.forfun.mmorpg.game.battle.model.BattleContext;
import org.forfun.mmorpg.game.database.converter.JpaObjectConverter;
import org.forfun.mmorpg.game.database.user.BaseEntity;
import org.forfun.mmorpg.game.database.user.dao.PlayerDao;
import org.forfun.mmorpg.game.scene.actor.ActorType;
import org.forfun.mmorpg.game.scene.actor.Creature;
import org.forfun.mmorpg.game.vip.model.VipRight;
import org.hibernate.annotations.Proxy;
import org.springframework.data.repository.CrudRepository;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;

@Entity
@Table(name = "playerent")
@Proxy(lazy = false)
public class PlayerEnt extends Creature implements BaseEntity<Long> {

    @Id
    @Column
    private long id;

    /**
     * 所属账号id
     */
    @Column
    private long accountId;

    @Column
    private String name;

    @Column
    private int level;

    @Column
    @Convert(converter = JpaObjectConverter.class)
    private VipRight vipRight;

    @Transient
    private BattleContext battleContext;

    @Override
    public CrudRepository<PlayerEnt, Long> getCrudRepository() {
        return GameContext.getBean(PlayerDao.class);
    }

    @Override
    @SuppressWarnings("unchecked")
    public Long getId() {
        return id;
    }

    public PlayerEnt() {

    }

    @Override
    public ActorType getType() {
        return ActorType.Player;
    }

    public boolean isOpenedFunction(int funcId) {
        return false;
    }

    public void setId(long id) {
        this.id = id;
    }

    public long getAccountId() {
        return accountId;
    }

    public void setAccountId(long accountId) {
        this.accountId = accountId;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getLevel() {
        return level;
    }

    public void setLevel(int level) {
        this.level = level;
    }

    public VipRight getVipRight() {
        return vipRight;
    }

    public void setVipRight(VipRight vipRight) {
        this.vipRight = vipRight;
    }

    public BattleContext getBattleContext() {
        return battleContext;
    }

    public void setBattleContext(BattleContext battleContext) {
        this.battleContext = battleContext;
    }

}

File: game-server/src/main/kotlin/org/forfun/mmorpg/game/database/user/entity/PlayerEnt.kt
package org.forfun.mmorpg.game.database.user.entity

import jakarta.persistence.*
import org.forfun.mmmorpg.game.scene.actor.Creature
import org.forfun.mmorpg.game.attribute.AttributeContainer
import org.forfun.mmorpg.game.base.GameContext
import org.forfun.mmorpg.game.buff.model.BuffContainer
import org.forfun.mmorpg.game.database.converter.JpaObjectConverter
import org.forfun.mmorpg.game.database.user.BaseEntity
import org.forfun.mmorpg.game.database.user.dao.PlayerDao
import org.forfun.mmorpg.game.vip.model.VipRight
import org.forfun.mmorpg.game.scene.actor.ActorType
import org.forfun.mmorpg.game.scene.model.Life
import org.hibernate.annotations.Proxy
import org.springframework.data.repository.CrudRepository

@Entity
@Table(name = "playerent")
@Proxy(lazy = false)
// 默认采用 FIELD 访问，确保 vipRight 上的 @field:Convert 生效（否则 Hibernate 会把
// VipRight 当成原生列类型，抛 Could not determine recommended JdbcType）
@Access(AccessType.FIELD)
class PlayerEnt() : Creature(), BaseEntity<Long> {

    init {
        buffContainer = BuffContainer()
        attrContainer = AttributeContainer()
    }

    fun setLife(health: Long, mana: Long) {
        life = Life(health, mana)
    }

    /**
     * 所属账号id
     */
    @field:Column
    var accountId: Long = 0

    @field:Column
    var name: String? = null

    @field:Column
    var level: Int = 0

    @field:Column
    private var _id: Long = 0

    @field:Column
    @field:Convert(converter = JpaObjectConverter::class)
    var vipRight: VipRight? = null


    @Transient
    override fun getCrudRepository(): CrudRepository<PlayerEnt, Long> {
        return GameContext.getBean(PlayerDao::class.java)
    }

    @Id
    @Column
    @Access(AccessType.PROPERTY)
    override fun getId(): Long {
        return _id
    }

    fun setId(id: Long) {
        _id = id
    }


    override val type: ActorType = ActorType.Player

    fun isOpenedFunction(funcId: Int): Boolean {
        return false
    }
}
